### PR TITLE
Enable ledger-api-test-tool to be used as a library

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
This is the third of four PRs in which https://github.com/digital-asset/daml/commit/6ea70c4b45d55cbf62caa3cc03dd70310ca79b87
has been broken up to facilitate review.

The endgame is to have the non-repudiation prototype merged. The
Ledger API test tool will be used to check that the proxy can be
safely put in front of a conformant ledger implementation without
causing conformance tests to not pass. The ability to use is as
a library facilitates the process, ensuring we don't have to set
up three running processes in Bazel (the participant, the proxy
and the test tool).

Part of #8598 

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by SamirTalwar-DA** [comment]
Could you explain a bit how you plan on using this?

Or better yet, write a test case that uses it? :upside_down_face:

**Comment by stefanobaghino-da** [comment]
> Could you explain a bit how you plan on using this?
> 
> Or better yet, write a test case that uses it?

As this is simply a refactoring, I would prefer to keep the PRs smalls and isolated to not overwhelm reviewers. The follow-up PR is [here](https://github.com/digital-asset/daml/pull/8608).

**Comment by stefanobaghino-da** [comment]
> Could you explain a bit how you plan on using this?
> 
> Or better yet, write a test case that uses it?

By the way, it's explained in the PR description.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala [hunk 1]
```diff
@@ -174,22 +176,60 @@ object LedgerApiTestTool {
...
+          uploadDars = config.uploadDars,
+          identifierSuffix = identifierSuffix,
+        )
     )
+  }
+
+  private def initializeParticipantChannels(
+      participants: Vector[(String, Int)],
+      tlsConfig: Option[TlsConfiguration],
+  )(implicit executionContext: ExecutionContext): Resource[ExecutionContext, Vector[Channel]] = {
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:214]: maybe worth factoring out the code for a single participant? it seems a bit odd to only have the version that accepts a `Vector` if it treats all of them independently.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala [hunk 2]
```diff
@@ -4,15 +4,15 @@
...
+import io.grpc.ManagedChannelBuilder
 
 private[testtool] final class LedgerSessionConfiguration(
-    participantAddresses: Vector[(String, Int)],
-    ssl: Option[TlsConfiguration],
+    participantChannelBuilders: Vector[ManagedChannelBuilder[_]],
     partyAllocation: PartyAllocationConfiguration,
     val shuffleParticipants: Boolean,
 ) {
+  participantChannelBuilders.foreach(println)
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala:14]: leftover debug output?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala [hunk 3]
```diff
@@ -140,65 +148,83 @@ final class LedgerTestCasesRunner(
...
+  private def prepareResourcesAndRun(implicit
+      executionContext: ExecutionContext
+  ): Future[Vector[LedgerTestSummary]] = {
+
+    val materializerResources =
+      ResourceOwner.forMaterializerDirectly(createActorSystem).acquire()
+
+    // When running the tests, explicitly use the materializer's execution context
+    // The tests will then be executed on it instead of the implicit one -- which
+    // should only be used to
```
Comment (cocreature) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala:213]: Seems like there is something missing in this sentence?

**Review by cocreature — APPROVED** [review]
Looks very sensible, thanks!

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala:213]: It's definitely not missing suspense.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala:14]: Definitely, thanks.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:214]: Why not.

Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/LedgerApiTestTool.scala:232]: Addressed by https://github.com/digital-asset/daml/pull/8607/commits/3093123330cfb7a6db61532d43ebeed33bb6c144

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerSessionConfiguration.scala]: Addressed by https://github.com/digital-asset/daml/pull/8607/commits/66344f5593d2d8a9513319a1c9fcaa47e33cdbdf

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/infrastructure/LedgerTestCasesRunner.scala]: Addressed by https://github.com/digital-asset/daml/pull/8607/commits/ef73c3d4fe8571375406332aa82dc0f2d42496da

**Pull request merged**